Add a readme (#35)
cilium/cilium-olm: README.md

The operator is configured using a [`CiliumConfig` CRD](config/crd/cilium.io_cilumconfigs.yaml). The operator watches a single `CiliumConfig` CR in `cilium` namespace. The CRD is a simple object that hold Helm values.

Default CR specs for each release are sourced from top-level manifest (e.g. [`ciliumconfig.v1.10.yaml`](ciliumconfig.v1.10.yaml)), a copy is made for each release in `manifests/` (e.g. [`manifests/cilium.v1.9.6/cluster-network-07-cilium-ciliumconfig.yaml`](manifests/cilium.v1.9.6/cluster-network-07-cilium-ciliumconfig.yaml).

### Cofiguration source code 

Since configuration for OLM is fairly complex to manage in static YAML format, it was written in [CUE](http://cuelang.org/).
CUE is not very easy to use on its own, so a simple opinionated utility called [`kg`][kg] is being used to execute CUE templates and write out YAML files.

The CUE templates can be found in [`config/operator/`]. `kg` is invoked with [a helper script](scripts/generate-configs.sh), which is typically driven by `make`, e.g. `make generate.configs.v1.9.6`.

### Generated configuration

Configuration that is generated from CUE templates will get used in variety of different ways:

> - contents of `mnanifests/` dir are used during installation in user guides as well as CI

Reviewer: ```suggestion
- contents of `manifests/` are used during installation in user guides as well as CI
```

> - contents of `bundles/` dir are used for building metadata bundle images

Reviewer: ```suggestion
- contents of `bundles/` are used for building metadata bundle images
```

### Images

The installation is reliant on operator application image, it is currently based on RedHat's helm-operator.

> For the purposes of certfication, the operator application image is accompanied by a metadata bundle image. Both images have to undergo certification tests in RedHat Partner Connect registry, maintainers should be able obtain access to this registry.

Reviewer 2: Links to the RedHat Partner Connect registry would be good

Author: I've added a section with links.

> For the purposes of certfication, the operator application image is accompanied by a metadata bundle image. Both images have to undergo certification tests in RedHat Partner Connect registry, maintainers should be able obtain access to this registry.

Reviewer: ```suggestion
For the purposes of certfication, the operator application image is accompanied by a metadata bundle image. Both images have to undergo certification tests in RedHat Partner Connect registry, maintainers should be able to obtain access to this registry.
```

The metadata bundle image contains just YAML files and no software as such, however it's a crucial part of OLM and when it comes to certification it's what is being tested the most. It's possible to install the operator based on this image, however it's not something that is recommended for Cilium, as Cilium should be installed during cluster bootstrap as otherwise users will have to carry out live network migration.

## Common Workflows

To work with this repository, a maintainer should have Docker with latest version of [`buildx` plugin](https://github.com/docker/buildx) installed, as well as Go toolchain.

To install `kg`, `imagine` and `opm` utilities to `GOPATH`, run:

```
(cd tools ; go install github.com/errordeveloper/imagine github.com/errordeveloper/kue/cmd/kg github.com/operator-framework/operator-registry/cmd/opm)
```

### Adding Cilium release

All releases and release candidates should be added to this repo for testing, albeit only latest stable release should be published.

To add a Cilium release, run:

```
scripts/add-release.sh 1.10.0
```

This will do the following:

> - create new dirs `{operator,manifests,bundles/cilium.v1.10.0` populate these with generated configs

Reviewer 2: ```suggestion
- create new dirs `{operator,manifests,bundles}/cilium.v1.10.0` populate these with generated configs
```

It can take some time, but often just a few hours. Sometimes the image doesn't show up in UI right away, you will need to check back every couple of hours.

Now it's time for the second step. Once operator application image has passed certification scan successfully and had been published, the metadata bundle can be pushed RedHat Partner Connect Registry also. To "promote" the metadata bundle run: 
```
scripts/push-to-scan-2-olm-metadata.sh 1.10.0
```

Next, access the project in RedHat Partner Connect portal and observe scan results. This often takes much longer, as it results in much more testing, i.e. a test cluster gets created to run the tests. By comparison it's not just a couple of hours, but more likely 6 hours or so.

If the certification tests pass, and if the bundle is for latest stable release and not a release candidate, it will need to published. In the UI it will show up as two images: `-cert` and `-mktpl`,  only `-cert` images should be published.

## Updating helm-operator base image

TODO

## Changing configuration

TODO

[okd-gsg]: https://docs.cilium.io/en/v1.10/gettingstarted/k8s-install-openshift-okd

> [olm]: https://docs.openshift.com/container-platform/4.7/operators/understanding/olm/olm-understanding-olm.html.

Reviewer: This link seems broken:

```suggestion
[olm]: https://docs.openshift.com/container-platform/4.7/operators/understanding/olm/olm-understanding-olm.html
```

- [Cilium OLM (bundle) in RedHat Catalog](https://catalog.redhat.com/software/containers/isovalent/cilium-olm-metadata/603fd17f69aea331dde395e4)

## Common Workflows

To work with this repository, a maintainer should have Docker with latest version of [`buildx` plugin](https://github.com/docker/buildx) installed, as well as Go toolchain.

To install `kg`, `imagine` and `opm` utilities to `GOPATH`, run:

```
(cd tools ; go install github.com/errordeveloper/imagine github.com/errordeveloper/kue/cmd/kg github.com/operator-framework/operator-registry/cmd/opm)
```

### Adding Cilium release

All releases and release candidates should be added to this repo for testing, albeit only latest stable release should be published in the RedHat Catalog.

To add a Cilium release, run:

```
scripts/add-release.sh 1.10.0
```

This will do the following:

- create new dirs `{operator,manifests,bundles}/cilium.v1.10.0` populate these with generated configs
- download Helm chart tarball and unpack it to 
- create a local commit that has all the changes that can be pushed to the repo

> Once the new commit is pushed, new images will be built in GitHub Actions.

Reviewer 3: I was trying to drive for instructing the reader not to do, rather than implying they must do something and then describing what happens when they do. Hence the "Push the commit". Otherwise it can be confusing whether the script above is responsible or somehow something else takes responsibility. Direct instructions mean it's very clear that the reader must do this thing and then something will happen.

> Once the new commit is pushed, new images will be built in GitHub Actions.

Author: I see, thanks for that! I've noticed previously that automated commits can be confusing to people in general.  Being extra explicit is always a good idea. Hopefully https://github.com/cilium/cilium-olm/commit/5656dad5fb4eaeaa51e88201997887935f4c2198 helps.

Next, images need to be copied from Quay to RedHat Partner Connect registry. This is a two-step process.

For this you will need to access RedHat Partner Connect registry and obtain credential for each of the image, which will need to be exported as following environment variables.

- `RHPC_PASSWORD_FOR_OLM_OPERATOR_IMAGE`
- `RHPC_PASSWORD_FOR_OLM_METADATA_IMAGE`
- `RHPC_USERNAME_FOR_PUBLISHED_IMAGES`
- `RHPC_PASSWORD_FOR_PUBLISHED_IMAGES`

The first step is to "promote" the operator application image from Quay to RedHat Partner Connect, for this run:

```
scripts/push-to-scan-1-olm-operator.sh 1.10.0
```

Now you will need to access the project in RedHat Partner Connect portal and observe scan results. Upon successful scan, you will need to publish the image (do not create `latest` tag when publishing). All releases and release candidates should be published at this stage, which is not the case for metdata bundles (see below).

It can take some time, but often just a few hours. Sometimes the image doesn't show up in UI right away, you will need to check back every couple of hours.

> Any scan failures should be logged as issues and addressed. With the operator image any fix will most likely result in a new image tag, so there is no need to worry about anything.

Author: @joestringer thanks for decrypting my own thinking here, it's funny how I avoided spelling out the quirky detail here... So there _were_ a few gotchas with the bundle image, so not the operator image. Basically there is no way to re-push or re-scan the same tag, all for some complicated reasons on RedHat end. I've previously used a trick where I appended a suffix to the tag by tweaking the Makefile... However, I think I have recently ironed all of these out by making sure that unique tag is generated for any changes to the bundle sources or template input parameters (see #43 & 3c8bb1e6254a9ab3cc1027ce49dba56531fc9125). That fix went in after this text was written, I'll make sure to rephrase this, good catch!

> Any scan failures should be logged as issues and addressed. With the operator image any fix will most likely result in a new image tag, so there is no need to worry about anything.

Reviewer 3: Ah thanks. I noticed the `[thing][]` syntax recently somewhere but I thought `[thing]` was enough in rST. TIL :)

[olm]: https://docs.openshift.com/container-platform/4.7/operators/understanding/olm/olm-understanding-olm.html
[kg]: https://github.com/errordeveloper/kue/blob/209ddfde99c57e533eae750aa7aaa16c0efeab04/cmd/kg/main.go

## Common Workflows

To work with this repository, a maintainer should have Docker with latest version of [`buildx` plugin](https://github.com/docker/buildx) installed, as well as Go toolchain.

To install `kg`, `imagine` and `opm` utilities to `GOPATH`, run:

```
(cd tools ; go install github.com/errordeveloper/imagine github.com/errordeveloper/kue/cmd/kg github.com/operator-framework/operator-registry/cmd/opm)
```

### Adding Cilium release

All releases and release candidates should be added to this repo for testing, albeit only latest stable release should be published in the RedHat Catalog.

To add a Cilium release, run:

```
scripts/add-release.sh 1.10.0
```

This will do the following:

- create new dirs `{operator,manifests,bundles}/cilium.v1.10.0` populate these with generated configs
- download Helm chart tarball and unpack it to 
- create a local commit that has all the changes that can be pushed to the repo

> Once the new commit is pushed, new images will be built in [GitHub Actions][].

Author: ```suggestion
Once changes are on the master branch of this repo, new images will be built in [GitHub Actions][].
```

For this you will need to access [RedHat Partner Connect][] registry and obtain credential for each of the images. Export the following environment variables using the credentials from RedHat Partner Connect:

- `export RHPC_PASSWORD_FOR_OLM_OPERATOR_IMAGE="_____"`
- `export RHPC_PASSWORD_FOR_OLM_METADATA_IMAGE="_____"`
- `export RHPC_USERNAME_FOR_PUBLISHED_IMAGES="_____"`
- `export RHPC_PASSWORD_FOR_PUBLISHED_IMAGES="_____"`

The first step is to "promote" the operator application image from Quay to RedHat Partner Connect, for this run:

```
scripts/push-to-scan-1-olm-operator.sh 1.10.0
```

Access OLM operator project in [RedHat Partner Connect][rhcp-projects] portal and observe the scan results for the newly pushed image. Upon successful scan, publish the image. Do not create `latest` tag when publishing.

All releases and release candidates should be published at this stage, but the metadata bundles will not yet be published as that will be handled in the second step below.

It can take some time, but often just a few hours. Sometimes the image doesn't show up in UI right away, you will need to check back every couple of hours.

> If an image scan fails, file an issue in this repository and address it issue. Once the chages are in master and [GitHub Actions][] build finished, re-run `scripts/push-to-scan-1-olm-operator.sh` and monitor the scan results.

Reviewer 3: typo nit:

```suggestion
If an image scan fails, file an issue in this repository and work to address the issue. Once the chages are in master and [GitHub Actions][] build finished, re-run `scripts/push-to-scan-1-olm-operator.sh` and monitor the scan results.
```
